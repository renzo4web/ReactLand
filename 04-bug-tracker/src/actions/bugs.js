import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { loadBugs } from '../helpers/loadBugs';
import { types } from '../types/types';

export const addNewBug = (bug) => {
    return async (dispatch, getState) => {
        const { name } = getState().auth;

        const newBug = {
            title: bug.name,
            reporter: name,
            description: bug.description,
            severity: bug.severity,
            assignee: bug.assignee || name,
            status: bug.status,
            createdAt: new Date().getTime(),
        };

        const docRef = await db.collection('bugs').add(newBug);
        try {
            const docAdded = await docRef;
            console.log(docAdded.id);
            await db
                .collection('bugs')
                .doc(docAdded.id)
                .update({ id: docAdded.id });
        } catch (error) {
            console.warn(error);
        }

        Swal.fire('Bug Added', '', 'success');
    };
};

export const startLoadingBugs = () => {
    return async (dispatch) => {
        const bugs = await loadBugs();
        dispatch(setBugs(bugs));
    };
};

export const setBugs = (bugs) => ({
    type: types.bugsLoad,
    payload: bugs,
});

export const startEditBug = (bug) => {
    return async (dispatch) => {
        await db.collection('bugs').doc(bug.id).update(bug);
        Swal.fire('Edited', 'the bug has been correctly edited', 'success');
    };
};

export const deleteBug = (id) => {
    return async (dispatch) => {
        await db.collection('bugs').doc(id).delete();
    };
};
