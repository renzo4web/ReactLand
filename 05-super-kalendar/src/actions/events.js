import toast from 'react-hot-toast';
import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

/*
    ADDING THE EVENTS TO DB
*/

export const eventStartAddNew = event => {
	return async (dispatch, getState) => {
		const { name, uid } = getState().auth;

		try {
			const response = await fetchWithToken('events', event, 'POST');
			const body = await response.json();

			if (body.ok) {
				dispatch(
					addNewEvent({
						...event,
						id: body.event.id,
						user: {
							_id: uid,
							name,
						},
					})
				);
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const addNewEvent = event => ({
	type: types.eventAddNew,
	payload: event,
});

export const setActiveEvent = event => ({
	type: types.eventSetActive,
	payload: event,
});

export const cleanActiveEvent = () => ({
	type: types.eventCleanActive,
});

/*
    UPDATE EVENT
*/
export const eventStartUpdate = event => {
	return async dispatch => {
		try {
			const response = await fetchWithToken(
				`events/${event.id}`,
				event,
				'PUT'
			);

			const body = await response.json();
			if (body.ok) {
				dispatch(updatedEvent(event));
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const updatedEvent = event => ({
	type: types.eventUpdated,
	payload: event,
});

/*
    DELETE EVENT
*/

export const eventStartDelete = () => {
	return async (dispatch, getState) => {
		const { activeEvent } = getState().calendar;
		try {
			const response = await fetchWithToken(
				`events/${activeEvent.id}`,
				{},
				'DELETE'
			);

			const { ok, msg } = await response.json();

			if (ok) {
				dispatch(deletedEvent());
			} else {
				toast.error(msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const deletedEvent = () => ({
	type: types.eventDeleted,
});

export const eventsCleanAll = () => ({
	type: types.eventCleanAll,
});

/*
    LOADING THE EVENTS FROM DB
*/

export const eventStartLoading = () => {
	return async dispatch => {
		try {
			const response = await fetchWithToken('events');

			const body = await response.json();

			if (body.ok) {
				const eventsPrepared = prepareEvents(body.events);
				dispatch(eventsLoaded(eventsPrepared));
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
		}
	};
};

const eventsLoaded = events => ({
	type: types.eventLoaded,
	payload: events,
});
