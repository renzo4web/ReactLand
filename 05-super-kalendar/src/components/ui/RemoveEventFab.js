import React from 'react';
import Button from 'react-bootstrap/Button';
import { RiDeleteBin2Line } from 'react-icons/ri';

const RemoveEventFab = props => {
	return (
		<Button
			{...props}
			className={
				'btn btn-danger rounded p-2 position-fixed bottom-0 start-0 m-4'
			}
		>
			<RiDeleteBin2Line size={45} />
		</Button>
	);
};

export default RemoveEventFab;
