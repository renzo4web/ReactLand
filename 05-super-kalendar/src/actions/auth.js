import { types } from '../types/types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import toast from 'react-hot-toast';
import { eventsCleanAll } from './events';

export const startLogin = (email, password) => {
	return async dispatch => {
		// http://localhost:4000/api/auth
		const response = await fetchWithoutToken(
			'auth',
			{ email, password },
			'POST'
		);
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startRegister = (name, email, password) => {
	return async dispatch => {
		const response = await fetchWithoutToken(
			'auth/new',
			{ email, password, name },
			'POST'
		);

		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startChecking = () => {
	return async dispatch => {
		const response = await fetchWithToken('auth/renew');
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			dispatch(checkingFinish());
		}
	};
};

const checkingFinish = () => ({
	type: types.authCheckingFinish,
});

export const startLogout = () => {
	return dispatch => {
		localStorage.clear();
		dispatch(logout());
		dispatch(eventsCleanAll());
	};
};

const logout = () => ({
	type: types.authLogout,
});

const login = user => ({
	type: types.authLogin,
	payload: user,
});
