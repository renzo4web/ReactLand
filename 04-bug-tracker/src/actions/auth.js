import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { bugsLogout } from './bugs';
import { finishLoading, startLoading } from './ui';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());

        firebase
            .auth()
            .signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((err) => {
                dispatch(finishLoading());
                Swal.fire('Error', err.message);
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase
            .auth()
            .signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                dispatch(login(uid, displayName));
            });
    };
};

export const startRegisterWithEmailPasswordName = (name, email, password) => {
    return (dispatch) => {
        firebase
            .auth()
            .createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                });

                dispatch(login(user.id, user.displayName));
            })
            .catch((err) => {
                Swal.fire('Error', err.message);
            });
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(bugsLogout());
        dispatch(logout());
    };
};

export const logout = () => ({
    type: types.logout,
});
