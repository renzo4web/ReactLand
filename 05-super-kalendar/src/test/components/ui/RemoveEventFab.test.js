import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { eventStartDelete } from '../../../actions/events';
import RemoveEventFab from '../../../components/ui/RemoveEventFab';

// Configuration to test components that use REDUX
jest.mock('../../../actions/events', () => ({
	eventStartDelete: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
// mock dispatch, because the action are already tested.
// just to know if the component are calling the dispatch
store.dispatch = jest.fn();

const wrapper = mount(
	<Provider store={store}>
		<RemoveEventFab onClick={() => store.dispatch(eventStartDelete())} />
	</Provider>
);

describe('Test components/ui/RemoveEventFab', () => {
	test('should render properly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should call eventStartDelete', () => {
		wrapper.find('button.btn').simulate('click');

		expect(eventStartDelete).toHaveBeenCalled();
	});
});
