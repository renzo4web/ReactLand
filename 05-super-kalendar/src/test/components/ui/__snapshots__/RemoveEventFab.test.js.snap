// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test components/ui/RemoveEventFab should render properly 1`] = `
<button
  className="btn btn-danger rounded p-2 position-fixed bottom-0 start-0 m-4 btn btn-primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <svg
    fill="currentColor"
    height={45}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width={45}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      key="0"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
        key="0"
      />
      <path
        d="M17 6h5v2h-2v13a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1V8H2V6h5V3a1 1 0 0 1 1-1h8a1 1 0 0 1 1 1v3zm1 2H6v12h12V8zm-4.586 6l1.768 1.768-1.414 1.414L12 15.414l-1.768 1.768-1.414-1.414L10.586 14l-1.768-1.768 1.414-1.414L12 12.586l1.768-1.768 1.414 1.414L13.414 14zM9 4v2h6V4H9z"
        key="1"
      />
    </g>
  </svg>
</button>
`;
